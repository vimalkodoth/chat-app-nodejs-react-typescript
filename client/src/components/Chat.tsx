/** @jsxImportSource @emotion/react */
import {
    Container,
    Form,
    Input,
    InputContainer,
    Main,
    MessageContainer,
    MessageContent,
    InfoMessageMeta,
    InfoMessageContent,
    MessageMeta,
    Sidebar,
    UsersHeaderStyles,
} from './Chat.styles';
import { useForm } from 'react-hook-form';
import { RootState } from '../redux/store';
import { useSelector } from 'react-redux';

export default function Chat({ children, ...resetProps }) {
    return (
        <div css={Container} {...resetProps}>
            {children}
        </div>
    );
}

Chat.Sidebar = function ChatSidebar({ children, ...resetProps }) {
    return <div css={Sidebar}>{children}</div>;
};

Chat.Main = function ChatMain({ children, messages }) {
    return <div css={Main}>{children(messages)}</div>;
};

export type TMessage = {
    room: string;
    nickname: string;
    time: string;
    message: string;
};
Chat.Input = function ChatInput({ onSelected }) {
    const {
        register,
        handleSubmit,
        formState: { errors },
        reset,
    } = useForm();

    const onSubmit = (data) => {
        const { message } = data;
        onSelected(message);
        reset({ message: '' });
    };

    return (
        <div css={Input}>
            <form css={Form} onSubmit={handleSubmit(onSubmit)}>
                <div css={InputContainer}>
                    <input
                        aria-label="message input"
                        id="message"
                        type="text"
                        {...register('message', {
                            required: true,
                        })}
                    />
                </div>
                <input type="submit" value="SEND" id="submit" />
            </form>
        </div>
    );
};

Chat.Label = function ChatDisplayName({
    header,
    value,
}: {
    header: string;
    value: string;
}) {
    return (
        <div>
            <h4 css={UsersHeaderStyles}>{header}</h4>
            <p>{value}</p>
        </div>
    );
};

Chat.User = function ChatUser() {
    const nickname = useSelector((state: RootState) => state.user.nickname);
    return (
        <div>
            <h4 css={UsersHeaderStyles}>User</h4>
            <p>{nickname}</p>
        </div>
    );
};

Chat.Users = function ChatUsers({ users }) {
    return (
        <>
            <h4 css={UsersHeaderStyles}>Online</h4>
            <ul>
                {users.map((user) => (
                    <li key={user.id}>{user.nickname}</li>
                ))}
            </ul>
        </>
    );
};

Chat.UserMessage = function ChatUserMessage({ data }) {
    const nickname = useSelector((state: RootState) => state.user.nickname);
    const isYou = nickname === data.nickname;
    const time = new Date(data.time);
    const hours = time.getHours().toString().padStart(2, '0');
    const minutes = time.getMinutes().toString().padStart(2, '0');
    return (
        <div css={MessageContainer(isYou)}>
            <div css={MessageContent(isYou)}>
                <p>{data.message}</p>
            </div>
            <div css={MessageMeta}>
                <p>
                    {hours}:{minutes} | {data.nickname}
                </p>
            </div>
        </div>
    );
};

Chat.InfoMessage = function ChatInfoMessage({ data }) {
    const time = new Date(data.time);
    const hours = time.getHours().toString().padStart(2, '0');
    const minutes = time.getMinutes().toString().padStart(2, '0');
    return (
        <div>
            <div css={InfoMessageContent}>
                <p>{data.message}</p>
            </div>
            <p css={InfoMessageMeta}>
                {hours}:{minutes}
            </p>
        </div>
    );
};
