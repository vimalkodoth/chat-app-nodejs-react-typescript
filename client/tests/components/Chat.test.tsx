import React from 'react';
import { shallow, mount, render } from 'enzyme';
import { act } from 'react-dom/test-utils';
import * as redux from 'react-redux';
import Chat from '../../src/components/common/Chat';

const testUser = 'testUser';
const spy = jest.spyOn(redux, 'useSelector');
spy.mockReturnValue({ nickname: 'testUser' });

describe('Chat', () => {
    it('should render Chat component', async () => {
        const wrapper = shallow(<Chat></Chat>);
        expect(wrapper.getElements()).toMatchSnapshot();
    });
    it('should render ChatSidebar component', async () => {
        const wrapper = shallow(<Chat.Sidebar></Chat.Sidebar>);
        expect(wrapper.getElements()).toMatchSnapshot();
    });
    it('should render ChatMain component', async () => {
        const wrapper = shallow(<Chat.Main></Chat.Main>);
        expect(wrapper.getElements()).toMatchSnapshot();
    });
    it('should render ChatLabel component', async () => {
        const wrapper = shallow(
            <Chat.Label header="header" value="value"></Chat.Label>
        );
        expect(wrapper.getElements()).toMatchSnapshot();
    });
    it('should render ChatUser component', async () => {
        const wrapper = shallow(<Chat.User />);
        expect(wrapper.getElements()).toMatchSnapshot();
    });
    it('should render ChatUsers component', async () => {
        const wrapper = shallow(
            <Chat.Users users={[{ id: 'id-0', nickname: testUser }]} />
        );
        expect(wrapper.getElements()).toMatchSnapshot();
    });
    it('should render ChatUserMessage component', async () => {
        const wrapper = shallow(
            <Chat.UserMessage
                data={{
                    time: new Date().toUTCString(),
                    message: 'hello!',
                    nickname: testUser,
                }}
            />
        );
        expect(wrapper.getElements()).toMatchSnapshot();
    });
    it('should render ChatUserMessage component', async () => {
        const wrapper = shallow(
            <Chat.InfoMessage
                data={{ time: new Date().toUTCString(), message: 'hello!' }}
            />
        );
        expect(wrapper.getElements()).toMatchSnapshot();
    });
    it('should render ChatInput component', async () => {
        const onSelected = jest.fn();
        const rendered = mount(<Chat.Input onSelected={onSelected} />);

        await act(async () => {
            rendered.find('input[id="message"]').instance().value =
                'Hey, a message';
            rendered.find('input[id="message"]').simulate('change');
            rendered.find('form').simulate('submit');
        });
        expect(rendered.getElements()).toMatchSnapshot();
        expect(onSelected).toHaveBeenCalled();
    });
});
