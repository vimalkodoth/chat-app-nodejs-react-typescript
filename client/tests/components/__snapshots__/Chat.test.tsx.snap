// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat should render Chat component 1`] = `
Array [
  <EmotionCssPropInternal
    __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Chat"
    __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
    css={
      Object {
        "map": undefined,
        "name": "1psxr7z",
        "next": undefined,
        "styles": "
    display: grid;
    box-sizing: border-box;
    grid-gap: 0.5em 0.5em;
    grid-template-columns: 1fr 2fr;
    grid-template-rows: 2fr 1fr;
    grid-template-areas:
        'sidebar main main'
        'sidebar main main'
        'sidebar main main'
        'input input input';
    border: 1px solid #e0e0e0;
    padding: 1em;
    min-height: 100%;
    max-height: 100%;
    & > div {
        border: 1px solid #eee;
        border-radius: 0.2em;
        padding: 0.5em;
        display: flex;
        flex-direction: column;
        overflow-y: scroll;
    }
",
        "toString": [Function],
      }
    }
  />,
]
`;

exports[`Chat should render ChatInput component 1`] = `
Array [
  <ChatInput
    onSelected={
      [MockFunction] {
        "calls": Array [
          Array [
            "Hey, a message",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  />,
]
`;

exports[`Chat should render ChatLabel component 1`] = `
Array [
  <div>
    <EmotionCssPropInternal
      __EMOTION_LABEL_PLEASE_DO_NOT_USE__="ChatDisplayName"
      __EMOTION_TYPE_PLEASE_DO_NOT_USE__="h4"
      css={
        Object {
          "map": undefined,
          "name": "18bu6t0",
          "next": undefined,
          "styles": "
    border-bottom: 1px solid #eee;
",
          "toString": [Function],
        }
      }
    >
      header
    </EmotionCssPropInternal>
    <p>
      value
    </p>
  </div>,
]
`;

exports[`Chat should render ChatMain component 1`] = `
Array [
  <EmotionCssPropInternal
    __EMOTION_LABEL_PLEASE_DO_NOT_USE__="ChatMain"
    __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
    css={
      Object {
        "map": undefined,
        "name": "1f702fh",
        "next": undefined,
        "styles": "
    grid-area: main;
",
        "toString": [Function],
      }
    }
  />,
]
`;

exports[`Chat should render ChatSidebar component 1`] = `
Array [
  <EmotionCssPropInternal
    __EMOTION_LABEL_PLEASE_DO_NOT_USE__="ChatSidebar"
    __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
    css={
      Object {
        "map": undefined,
        "name": "sh2e66",
        "next": undefined,
        "styles": "
    grid-area: sidebar;
",
        "toString": [Function],
      }
    }
  />,
]
`;

exports[`Chat should render ChatUser component 1`] = `
Array [
  <div>
    <EmotionCssPropInternal
      __EMOTION_LABEL_PLEASE_DO_NOT_USE__="ChatUser"
      __EMOTION_TYPE_PLEASE_DO_NOT_USE__="h4"
      css={
        Object {
          "map": undefined,
          "name": "18bu6t0",
          "next": undefined,
          "styles": "
    border-bottom: 1px solid #eee;
",
          "toString": [Function],
        }
      }
    >
      User
    </EmotionCssPropInternal>
    <p>
      Object {
        "nickname": "testUser",
      }
    </p>
  </div>,
]
`;

exports[`Chat should render ChatUserMessage component 1`] = `
Array [
  <EmotionCssPropInternal
    __EMOTION_LABEL_PLEASE_DO_NOT_USE__="ChatUserMessage"
    __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
    css={
      Object {
        "map": undefined,
        "name": "1mzvkn0",
        "next": undefined,
        "styles": "
    max-width: 50%;
    p {
        margin: 5px 0;
    }
    align-self: flex-start;
    
       align-self:flex-end;
    
",
        "toString": [Function],
      }
    }
  >
    <EmotionCssPropInternal
      __EMOTION_LABEL_PLEASE_DO_NOT_USE__="ChatUserMessage"
      __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
      css={
        Object {
          "map": undefined,
          "name": "41qawm",
          "next": undefined,
          "styles": "
    font-size: 1rem;
    background-color: #128c7e;
    
        background-color: #34B7F1;
    
    color: #fff;
    padding: 0.5em;
    border-radius: 0.6em;
    p {
        margin: 0;
    }
",
          "toString": [Function],
        }
      }
    >
      <p>
        hello!
      </p>
    </EmotionCssPropInternal>
    <EmotionCssPropInternal
      __EMOTION_LABEL_PLEASE_DO_NOT_USE__="ChatUserMessage"
      __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
      css={
        Object {
          "map": undefined,
          "name": "1r0slvy",
          "next": undefined,
          "styles": "
    font-size: 0.75rem;
    margin-top: 0.2em;
",
          "toString": [Function],
        }
      }
    >
      <p>
        21
        :
        49
         | 
        testUser
      </p>
    </EmotionCssPropInternal>
  </EmotionCssPropInternal>,
]
`;

exports[`Chat should render ChatUserMessage component 2`] = `
Array [
  <div>
    <EmotionCssPropInternal
      __EMOTION_LABEL_PLEASE_DO_NOT_USE__="ChatInfoMessage"
      __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
      css={
        Object {
          "map": undefined,
          "name": "w92gfy",
          "next": undefined,
          "styles": "
    p {
        margin-bottom: 3px;
    }
",
          "toString": [Function],
        }
      }
    >
      <p>
        hello!
      </p>
    </EmotionCssPropInternal>
    <EmotionCssPropInternal
      __EMOTION_LABEL_PLEASE_DO_NOT_USE__="ChatInfoMessage"
      __EMOTION_TYPE_PLEASE_DO_NOT_USE__="p"
      css={
        Object {
          "map": undefined,
          "name": "1k67m2o",
          "next": undefined,
          "styles": "
    font-size: 12px;
",
          "toString": [Function],
        }
      }
    >
      21
      :
      49
    </EmotionCssPropInternal>
  </div>,
]
`;

exports[`Chat should render ChatUsers component 1`] = `
Array [
  <React.Fragment>
    <EmotionCssPropInternal
      __EMOTION_LABEL_PLEASE_DO_NOT_USE__="ChatUsers"
      __EMOTION_TYPE_PLEASE_DO_NOT_USE__="h4"
      css={
        Object {
          "map": undefined,
          "name": "18bu6t0",
          "next": undefined,
          "styles": "
    border-bottom: 1px solid #eee;
",
          "toString": [Function],
        }
      }
    >
      Online
    </EmotionCssPropInternal>
    <ul>
      <li>
        testUser
      </li>
    </ul>
  </React.Fragment>,
]
`;
